### PR TITLE
Replace text angular editor with trix editor in About Us and Shopfront message fields

### PR DESCRIPTION
**Pull request by cillian**
#### What? Why?

- For https://github.com/openfoodfoundation/openfoodnetwork/issues/9481 and https://github.com/openfoodfoundation/openfoodnetwork/issues/9490 

The trix editor was added in https://github.com/openfoodfoundation/openfoodnetwork/pull/11140, this PR follows on from that and replaces the remaining instances of the text-angular editor.

#### What should we test?

- Test the Trix editor is working for the following fields...
  * Enterprise Settings > About > About us 
  * Enterprise Settings > Shop preferences > Shopfront message
  * Enterprise Settings > Shop preferences > Shopfront closed message
  * Enterprise Group Settings > About > About
 
#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

**Before**

about-us-before

shop-messages-before

groups-about-us-before

**After** 

about-us-after

shop-messages-after

groups-about-us-after

Note, in the above screenshots the trix fields are on new lines instead of being alongside the field label like they usually are. This is because there was a horizontal scrollbar if it was beside the field label, see:

shop-messages-after-with-horizontal-scollbar

I also added some wrapping to CSS to the trix editor button groups so they will automatically wrap and stop the horizontal scrollbar on smaller screens, although another solution might be to remove the whitespace to the right, see:

shop-messages-wrapping-after

### CODE REVIEW

File: app/models/enterprise.rb
Comment (cillian) [inline app/models/enterprise.rb:254]: I introduced a `sanitize_and_enforce_link_target_blank` method because I wasn't sure if  `target="_blank"`  is needed _all_ the time, or just _some_ of the time.

One other thing I wasn't sure about was whether to use [HtmlSanitizer](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/services/html_sanitizer.rb) or [TrixSanitizer](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/services/trix_sanitizer.rb), they both seem similar but they are not the exact same.

Comment (rioug) [reply app/models/enterprise.rb:254]: From memory `HtmlSanitizer` was introduced specifically for `Enterprise` model so this is the correct place to add `sanitize_and_enforce_link_target_blank`.

**Review by rioug — APPROVED** [review]
yeah less angular :broom:  ! Thanks @cillian :pray:

File: app/models/enterprise.rb
Comment (dacook) [reply app/models/enterprise.rb:254]: Sounds good to me. I'm guessing that the previous solution did the same thing? (enforce link `target="blank"` on _all_ links). 
I'm guessing it was because the links would normally point to an external website. It raises the question what if people use internal links to the current website. But that should be answered in a separate issue, if needed.

File: app/views/admin/enterprise_groups/_form_about.html.haml
Comment (dacook) [inline app/views/admin/enterprise_groups/_form_about.html.haml:4]: If the subject of the form `f` is `@enterprise_group`, then the ID and Value should be assigned automatically. Does it work if they are omitted?

```suggestion
    = f.hidden_field :long_description, id: "enterprise_group_long_description"
```

(Same for others below)

File: config/locales/en.yml
Comment (dacook) [inline config/locales/en.yml:3752]: Thanks for adding Trix translation keys. 

These two old labels seem equivalent to new labels (I haven't checked though)
* "Insert / edit link" -> "Link"
* "Please enter a URL to insert" -> "Enter a URL…"

If so, I think it would help to use the old labels. That way, it will probably be easy for translators to copy the existing labels when adding translations for Trix. Does that make sense?

**Review by dacook — COMMENTED** [review]
Great work, it looks much cleaner than before!
I like how you adapted to fit the extra Trix buttons in, and it means there's more space for the text area too. 

I have a couple of comments that might be worth trying. Do they make sense? 
If not, we can proceed as is.

File: app/models/enterprise.rb
Comment (cillian) [reply app/models/enterprise.rb:254]: Yes, previously the text angular editor would add a `target="_blank"` attribute every time you inserted a link, see https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/assets/javascripts/admin/utils/directives/textangular_links_target_blank.js.coffee

File: config/locales/en.yml
Comment (cillian) [reply config/locales/en.yml]: Good idea, that's updated in https://github.com/openfoodfoundation/openfoodnetwork/pull/12734/commits/2d6ffc0ca16459f0d4ee3abdaf82189ddd0a8656 to use the old "Please enter a URL to insert" label.

The "Link" translation is used in two places at once so the "Insert / edit link" label is a bit more awkward, see...

link-translation

File: app/webpacker/controllers/trixeditor_controller.js
Comment (cillian) [inline app/webpacker/controllers/trixeditor_controller.js:85]: ` Trix.config.lang[key] = translation;` only seems to work in Firefox for some reason. The translations have to be set more manually for Chrome.

File: config/locales/en_AU.yml
Comment (dacook) [inline config/locales/en_AU.yml:2977]: Actually, these will already have the default loaded from `en.yml`

In general , we need to avoid changing any locale files other than the default, as described here:
https://github.com/openfoodfoundation/openfoodnetwork/wiki/Internationalisation-(i18n)#development

I have to admit, I don't know if this commit will be a problem or not, but I know it's not necessary so perhaps it's safest to remove it.

File: app/webpacker/controllers/trixeditor_controller.js
Comment (dacook) [reply app/webpacker/controllers/trixeditor_controller.js]: That's interesting, and unfortunate because you have map quite a lot data.

According to this wiki, it suggests overwriting the `lang` config in one command.
> `Trix.config.lang = translation`
https://github.com/basecamp/trix/wiki/I18n

But you tried that and it didn't work, so it would be worth raising an issue.
I had a look for issues and could only find one. 
The final comment shows another method using `Object.assign`. I don't know if that would make a different, but perhaps it's worth trying.
https://github.com/basecamp/trix/issues/874#issuecomment-866693637

File: config/locales/ar.yml
Comment (dacook) [inline config/locales/ar.yml:3329]: Sorry I just meant to update it in `en.yml`. In general we should avoid changing other locale files directly. I'm unsure if adding new keys is a problem though.

Comment (dacook) [reply config/locales/ar.yml:3329]: @mkllnk can you help with a query here? 
For convenience, Cillian has copied an existing translation to a new key (the existing translation was for angular editor, and there is an equivalent in the new trix editor). Do you know if it's ok, and if it's helpful to add new keys to the locale files?
Or should we remove this before merging?

File: config/locales/ar.yml
Comment (mkllnk) [inline config/locales/ar.yml:1055]: Please remove all changes to locales other than en.yml.

File: config/locales/ar.yml
Comment (mkllnk) [reply config/locales/ar.yml:3329]: We should remove this addition. Transifex will handle it. It recognises existing translations.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Great work! Thank you.

We just need to revert he changes to the other locales. Otherwise this is great!

File: config/locales/ar.yml
Comment (cillian) [reply config/locales/ar.yml]: :+1: Fixed by https://github.com/openfoodfoundation/openfoodnetwork/pull/12734/commits/dd5175558e7123052e900ea00b93d29df8a2cc4f

File: app/webpacker/controllers/trixeditor_controller.js
Comment (cillian) [reply app/webpacker/controllers/trixeditor_controller.js]: Good find, fixed by https://github.com/openfoodfoundation/openfoodnetwork/pull/12734/commits/63c62cae08ec22fe234ae0e1b8c3dcde4ac30bec

**Comment by cillian** [comment]
Thanks. I have reverted those changes to non `en` locale files now and simplified how the Trix translations are set, let me know if you want me to squash the commits.

**Review by rioug — APPROVED** [review]

**Review by dacook — APPROVED** [review]
Great, all ready to test!

**Comment by drummer83** [comment]
@cillian Github is saying there are changes requested. Could you please have a look? I assume it's been addressed already?

**Review by mkllnk — APPROVED** [review]
> Github is saying there are changes requested.

Github was waiting for my review. But the other approvals should have been enough. Anyway, you get may approval, too. :smile:

**Comment by drummer83** [comment]
Hi @cillian
Sorry, there's a conflict here. Could you please have a look?
Thanks!

**Comment by drummer83** [comment]
For future testing I have documented the "before" state already.

- Already existing Trix editor for White Label Custom Tab Content:
image

- Same on narrow screens (scroll bar and still truncated buttons):
image

- Old editor for Home and Shopfront Closed message with test content using all available options:
image

- What it looks like in the shopfront:
image

**Comment by cillian** [comment]
Thanks @drummer83 I have fixed that conflict now. 

For reference, it was this...

```
  # Remove any unsupported HTML.
<<<<<<< HEAD
  def long_description
    HtmlSanitizer.sanitize_and_enforce_link_target_blank(super)
  end

  # Remove any unsupported HTML.
=======
>>>>>>> master
  def long_description=(html)
    super(HtmlSanitizer.sanitize_and_enforce_link_target_blank(html))
  end
```

To fix it I removed the `#long_description` method like what was done in https://github.com/openfoodfoundation/openfoodnetwork/commit/d061fe8ad926e60294b68b9c5dfdfc4d37a20c5b because the enforcing of `target="_blank"` will be added to links when assigning the `long_description=` attribute. Before this PR all links already had `target="_blank"` on them because that was the default behaviour of the old angular text editor.

**Comment by drummer83** [comment]
Hi @cillian,
Thanks for working on this one! Here are my test results after staging.

## After staging the PR

- Shopfront display remains unchanged.
- Back office editor is updated and styling is adapted to newly available options (One header only, no underline, ...).
grafik

## After staging the PR and updating the content
- Shopfront display is adapted to use newly available options.
grafik
- Behavior on narrow screens is much better now.
grafik

I find it surprising to see, that it's enough to edit any of the content to have the new styles applied to all of the occurences. It's no problem, maybe even an advantage looking at consistency. I assume this happens because when clicking Update, all of the tabs are saved and not just the tab I'm looking at. 

## Summary
- Styling changes in the shop front due to different styling options in the new editor.
- This is not critical, because all information is still available.
- Small screens are much better supported.
- No problems found.

We can merge this one, but the branch is out of date and I can't rebase due to a conflict. I will move this to Ready To Go but add the feedback needed label to have someone else have a look at it. Maybe @filipefurtad0 or @sigmundpetersen could have a look? 🙏 

Thanks again Cillian! 🚀

**Comment by sigmundpetersen** [comment]
Merging is blocked because reapproval is needed after Cillian resolved the last conflict.
Maybe @mkllnk or @rioug could have a quick look when coming back to work and merge if everything looks Ok 👍

**Review by rioug — APPROVED** [review]

**Pull request merged**
